### PR TITLE
Fix increase by percentage returns 1 if experience or max_experience are equal to zero

### PR DESCRIPTION
**Pull request by newfrenchy83**
Resolve https://github.com/wesnoth/wesnoth/issues/8374 issue

### CODE REVIEW

File: src/units/unit.cpp [hunk 1]
```diff
@@ -2214,7 +2214,10 @@ void unit::apply_builtin_effect(std::string apply_to, const config& effect)
 		}
 
 		if(increase.empty() == false) {
-			experience_ = utils::apply_modifier(experience_, increase, 1);
```
Comment (stevecotton) [inline src/units/unit.cpp:2217]: The third argument is a minimum value; setting it to zero fixes the bug.

@CelticMinstrel can you remember why it was set to 1 in #618?
Having a minimum of 1 for max_experience makes sense to me, so maybe it was copied from that.

Comment (CelticMinstrel) [reply src/units/unit.cpp:2217]: I don't remember, no.

Comment (newfrenchy83) [reply src/units/unit.cpp:2217]: > The third argument is a minimum value; setting it to zero fixes the bug.
> 
fixed

**Review by stevecotton — APPROVED** [review]
Looks good to me.

@CelticMinstrel I think the only question is whether this goes in 1.18, or (as the buggy behavior was already in 1.14 and 1.16) gets postponed until 1.19. I'm 95% sure this should go in 1.18.

**Comment by newfrenchy83** [comment]
> Looks good to me.
> 
> @CelticMinstrel I think the only question is whether this goes in 1.18, or (as the buggy behavior was already in 1.14 and 1.16) gets postponed until 1.19. I'm 95% sure this should go in 1.18.

I think for 1.18 also

File: data/test/scenarios/wml_tests/UnitsWML/increase_experience_by_percentage.cfg [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# wmllint: no translatables
+
+#####
+# API(s) being tested: [unit]experience=0
```
Comment (stevecotton) [inline data/test/scenarios/wml_tests/UnitsWML/increase_experience_by_percentage.cfg:4]: Should be `[effect]apply_to=experience`

Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/increase_experience_by_percentage.cfg]: corected

**Comment by newfrenchy83** [comment]
@CelticMinstrel ,@stevecotton  i see what utils::apply_modifier() with 1by default is used in attack_type.cpp(attack_type::apply_modification(), it is intended or some modification should be to zero?

**Review by stevecotton — APPROVED** [review]
Will merge when the CI finishes.

**Comment by stevecotton** [comment]
> @CelticMinstrel ,@stevecotton i see what utils::apply_modifier() with 1by default is used in attack_type.cpp(attack_type::apply_modification(), it is intended or some modification should be to zero?

Having a minimum of 1 for number of strikes and attack cost both seem correct to me.
Having a minimum of 1 for movement used sounds odd; not unreasonable but also IIUC not the same as the disengage ability would have. Please don't change it in this PR though.

**Pull request merged**